src/plan/aggregate.rs

```diff
@@ -203,11 +224,149 @@ impl<P: Implementable> Implementable for Aggregate<P> {
             //             let sum_square = diff_pair.element1.element1 as f64;
             //             let sum = diff_pair.element1.element2 as f64;
             //             let c = diff_pair.element2 as f64;
-            //
             //             v.push(Value::Rational32(Rational32::from_float((sum_square/c - (sum/c).powi(2)).sqrt()).unwrap()));
             //             v
             //         }),
             // },
         }
     }
 }
+
+/// [WIP]
+/// A plan stage applying multiple aggregations to
+/// bindings for the specified symbols. Very WIP.
+#[derive(Deserialize, Clone, Debug)]
+pub struct AggregateMulti<P: Implementable> {
+    /// TODO
+    pub variables: Vec<Var>,
+    /// Plan for the data source.
+    pub plan: Box<P>,
+    /// Logical predicate to apply.
+    pub aggregation_fn: Vec<AggregationFn>,
+    /// Relation symbols that determine the grouping.
+    pub key_symbols: Vec<Var>,
+    /// Aggregation symbols
+    pub aggregation_symbols: Vec<Var>,
+}
+
+impl<P: Implementable> Implementable for AggregateMulti<P> {
+    fn implement<'a, 'b, A: Allocate>(
+        &self,
+        nested: &mut Iterative<'b, Child<'a, Worker<A>, u64>, u64>,
+        local_arrangements: &RelationMap<Iterative<'b, Child<'a, Worker<A>, u64>, u64>>,
+        global_arrangements: &mut QueryMap<isize>,
+    ) -> SimpleRelation<'b, Child<'a, Worker<A>, u64>> {
+        let relation = self.plan
+            .implement(nested, local_arrangements, global_arrangements);
+
+        let mut value_offsets = Vec::new();
+        let mut seen = Vec::new();
+
+        for sym in self.aggregation_symbols.iter() {
+            if !seen.contains(&sym) {
+                seen.push(&sym);
+                value_offsets.push(seen.iter().position(|&v| sym == v).unwrap());
+            } else {
+                value_offsets.push(seen.iter().position(|&v| sym == v).unwrap());
+            }
+        }
+
+        let mut result_position = self.key_symbols.clone();
+        result_position.append(&mut self.aggregation_symbols.clone());
+
+        let mut variables = self.variables.clone();
+
+        let mut sym_position = Vec::new();
+
+        for sym in result_position.iter() {
+            let i = variables.iter().position(|&v| *sym == v).unwrap();
+            sym_position.push(i);
+            variables[i] = 0;
+        }
+
+        let tuples = relation.tuples_by_symbols(&self.key_symbols);
+
+        let aggregation_fn = self.aggregation_fn.clone();
+
+        SimpleRelation {
+            symbols: self.variables.to_vec(),
+            tuples: tuples
+                .group(move |_key, vals, output| {
+                    let mut results = Vec::new();
+                    for (i, agg_fn) in aggregation_fn.iter().enumerate() {
+                        let mut values: Vec<Value> =
```

Comment:
Ok so this is a bit scary indeed. Let's talk about this and the alternatives in person. If it works I'm ok with merging it for now, but we should spend some more time thinking about this.

Reply:
Sure sounds good !